Nama : Rika Ajeng Finatih

NIM : 121450036

Kelas : Bioinformatika RB

## **LATIHAN**

1. Silahkan buat analisis sekuen dengan ID: NC_045512:

    a. Temukan Key dan Value dari ID tersebut

    b. Jelaskan informasi terkait sekuen dengan ID tersebut

2. Downloadlah dataset variant dari covid-19 informasi ada disini https://en.wikipedia.org/wiki/Variants_of_SARS-CoV-2:

    a. Downloadlah menggunakan file referensi dengan format .txt

    b. Gunakan fungsi untuk melakukan merger file dengan format **(fasta dan ganbank)**


3. Tugas dikumpulkan dalam format nim_nama.ipynb (untuk kodingan) dan nim_nama.fasta & nim_nama.gb pada GCR sebelum 1 Oktober 2024, pukul 14.30

## **Nomor 1**

Sekuen `NC_045512` merupakan salah satu sekuen genom dari virus SARS-CoV-2 (COVID 19) yang berisi referensi genom virus yang digunakan dalam berbagai studi tentang virus tersebut. Biasanya file sekuen ini dalam format GenBank atau FASTA yang berisi informasi seperti kode DNA/RNA, gen, protein, dan anotasi terkait.

Link untuk mengunduh: [NCBI - Sekuen NC_045512](https://www.ncbi.nlm.nih.gov/nuccore/NC_045512)

In [ ]:
# Install Biopython
!pip install biopython

In [ ]:
from Bio import SeqIO  # Mengimpor library Biopython untuk membaca file sekuen

# Menentukan jalur file GenBank yang berisi sekuen NC_045512 (file ini harus sudah diunduh dari NCBI)
file_path = "/content/121450036_Rika Ajeng Finatih.gb"

# Membuka file GenBank dan membaca sekuen
with open(file_path, "r") as file:
    record = SeqIO.read(file, "genbank")  # Membaca file GenBank

# Menampilkan ID sekuen
print(f"ID: {record.id}")

# Menampilkan deskripsi dari sekuen
print(f"Deskripsi: {record.description}")

# Menampilkan panjang sekuen dalam satuan basa (nukleotida)
print(f"Panjang Sekuen: {len(record.seq)} basa")

# Menampilkan 100 huruf pertama dari sekuen sebagai contoh
print(f"Contoh Sekuen: {record.seq[:100]}")

# Menampilkan anotasi (informasi tambahan tentang sekuen)
print("Anotasi: ", record.annotations)

# Mendapatkan fitur gen dari sekuen yang ada (fitur gen mencakup gen yang ditemukan di sekuen)
gene_features = [feature for feature in record.features if feature.type == "gene"]

ID: NC_045512.2
Deskripsi: Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome
Panjang Sekuen: 29903 basa
Contoh Sekuen: ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTC
Anotasi:  {'molecule_type': 'ss-RNA', 'topology': 'linear', 'data_file_division': 'VRL', 'date': '18-JUL-2020', 'accessions': ['NC_045512'], 'sequence_version': 2, 'gi': '1798174254', 'keywords': ['RefSeq'], 'source': 'Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)', 'organism': 'Severe acute respiratory syndrome coronavirus 2', 'taxonomy': ['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus', 'Severe acute respiratory syndrome-related coronavirus'], 'references': [Reference(title='A new coronavirus associated with human respiratory disease in China', ...), Reference(title='Programmed ribosomal

Sekuen NC_045512.2 adalah sekuen genom lengkap dari virus *Severe Acute Respiratory Syndrome Coronavirus 2 (SARS-CoV-2)*, yang pertama kali diisolasi dari pasien di Wuhan, Tiongkok, dan dikenal sebagai penyebab penyakit COVID-19. Sekuen ini memiliki panjang 29.903 basa nukleotida, yang mencakup seluruh informasi genetik yang diperlukan bagi virus untuk bereplikasi di dalam sel inangnya. Sekuen ini termasuk dalam kelompok virus RNA untai tunggal (ss-RNA), yang berarti bahwa genomnya berfungsi langsung sebagai template untuk sintesis protein virus. Dengan tipe topologi linear, genom ini dibedakan dari beberapa jenis virus lain yang memiliki genom melingkar. Kode akses NC_045512 mengidentifikasi versi sekuen ini dalam sistem NCBI RefSeq, sebuah database referensi yang dikurasi dan dianggap sebagai acuan. Versi sekuen ini telah diperbarui pada 18 Juli 2020, yang merupakan versi kedua setelah revisi sebelumnya.

Dari segi klasifikasi taksonomi, SARS-CoV-2 berada dalam ordo Nidovirales dan termasuk dalam famili Coronaviridae serta genus Betacoronavirus, yang juga mencakup virus SARS pertama. Anotasi taksonomi ini menempatkan virus ini dalam konteks evolusioner bersama virus-virus RNA lain. Beberapa referensi penelitian menyertai sekuen ini, di antaranya publikasi yang menggambarkan penemuan awal virus di Wuhan serta beberapa penelitian lain yang fokus pada mekanisme translasi dan replikasi RNA virus. Informasi tambahan terkait perakitan genom menunjukkan bahwa metode Megahit dengan teknologi Illumina digunakan untuk melakukan sekuensing pada genom ini.

In [ ]:
# Menampilkan jumlah fitur gen yang ditemukan
print(f"Jumlah Fitur Gen: {len(gene_features)}")

# Menampilkan informasi tentang gen pertama yang ada dalam sekuen
if len(gene_features) > 0:  # Memastikan ada gen yang ditemukan
    first_gene = gene_features[0]
    print(f"Lokasi Gen Pertama: {first_gene.location}")  # Menampilkan lokasi gen dalam sekuen
    print(f"Kualifikasi Gen Pertama: {first_gene.qualifiers}")  # Menampilkan kualifikasi (metadata) tentang gen pertama
else:
    print("Tidak ada fitur gen yang ditemukan.")


Jumlah Fitur Gen: 11
Lokasi Gen Pertama: [265:21555](+)
Kualifikasi Gen Pertama: {'gene': ['ORF1ab'], 'locus_tag': ['GU280_gp01'], 'db_xref': ['GeneID:43740578']}


Genom SARS-CoV-2 terdiri dari beberapa fitur genetik yang berperan penting dalam fungsionalitas dan replikasi virus. Dalam kasus ini, terdapat 11 fitur gen yang telah diidentifikasi. Gen pertama terletak pada posisi [265:21555] dalam urutan genom, yang menunjukkan bahwa gen ini membentang dari nukleotida ke-265 hingga nukleotida ke-21.555 pada untai positif (+). Gen ini dikenal sebagai ORF1ab (Open Reading Frame 1ab), yang merupakan gen terbesar dan salah satu yang paling penting pada genom virus SARS-CoV-2.

Gen ORF1ab bertanggung jawab untuk menghasilkan protein non-struktural melalui mekanisme yang dikenal sebagai pemrosesan proteolitik. Protein non-struktural yang dihasilkan oleh gen ini terlibat dalam replikasi dan transkripsi RNA virus. Salah satu fungsi utamanya adalah sebagai bagian dari kompleks replikasi-transkripsi virus, yang penting dalam siklus hidup SARS-CoV-2.

Dalam anotasi ini, gen ORF1ab juga dikaitkan dengan locus_tag yang diberi kode GU280_gp01 serta referensi database eksternal GeneID:43740578. Ini berarti bahwa gen tersebut telah diidentifikasi dan didokumentasikan di berbagai basis data genomik, termasuk di database NCBI Gene, yang menyediakan informasi tambahan mengenai fungsi biologis gen tersebut.

Secara keseluruhan, informasi ini memberikan gambaran mendetail tentang gen pertama pada genom SARS-CoV-2 dan peran pentingnya dalam replikasi virus serta keterkaitannya dengan referensi eksternal dalam penelitian ilmiah lebih lanjut.








## **Nomor 2**

Akses halaman Wikipedia tentang varian COVID-19: [Variants_of_SARS-CoV-2](https://en.wikipedia.org/wiki/Variants_of_SARS-CoV-2)

In [ ]:
import requests

def unduh_sekuens_batch(urls, ukuran_batch=200):
    """Fungsi untuk mengunduh sekuens dalam batch."""
    for i in range(0, len(urls), ukuran_batch):
        batch_urls = urls[i:i + ukuran_batch]
        for idx, url in enumerate(batch_urls, start=i):
            response = requests.get(url)

            if response.status_code == 200:
                # Mengubah ekstensi file menjadi .fasta
                nama_file = f"sekuens_batch_{idx}.fasta"
                with open(nama_file, "wb") as f:
                    f.write(response.content)
                print(f"Berhasil mengunduh sekuens {idx + 1} dari {url}")
            else:
                print(f"Gagal mengunduh sekuens {idx + 1} dari {url}")

# Contoh penggunaan dengan daftar URL
urls = ["https://www.google.com/url?q=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FVariants_of_SARS-CoV-2"]

unduh_sekuens_batch(urls, ukuran_batch=200)


Berhasil mengunduh sekuens 1 dari https://www.google.com/url?q=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FVariants_of_SARS-CoV-2


In [ ]:
from Bio import SeqIO

# Fungsi untuk membaca sekuens dari file FASTA
def baca_fasta(jalur_file):
    sekuens = list(SeqIO.parse(jalur_file, "fasta"))
    print(f"Ditemukan {len(sekuens)} sekuens dalam {jalur_file}")

    for seq_record in sekuens[:5]:  # Tampilkan 5 sekuens pertama
        print(f"ID: {seq_record.id}")
        print(f"Sekuen: {seq_record.seq[:100]}...")  # Tampilkan 100 basis pertama

    return sekuens

# Contoh penggunaan
fasta_sequences = baca_fasta("sequences_batch_200.fasta")


Ditemukan 200 sekuens dalam sequences_batch_200.fasta
ID: PQ393194.1
Sekuen: AAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTA...
ID: PQ393193.1
Sekuen: GCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGCAGGC...
ID: PQ393192.1
Sekuen: AAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTA...
ID: PQ393191.1
Sekuen: TCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTTTGTCCGGGT...
ID: PQ393190.1
Sekuen: AAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTA...


In [ ]:
from Bio import SeqIO

# Daftar file FASTA yang ingin digabungkan
daftar_fasta = ["sequences_batch_0.fasta", "sequences_batch_200.fasta",
                "sequences_batch_400.fasta", "sequences_batch_600.fasta",
                "sequences_batch_800.fasta"]

# Nama file output
file_output = "121450036_Rika Ajeng Finatih.fasta"

# Membuka file output untuk ditulis
with open(file_output, "w") as outfile:
    for file_fasta in daftar_fasta:
        # Memparse setiap file FASTA input
        with open(file_fasta, "r") as infile:
            # Menggunakan SeqIO.parse untuk membaca sekuens dan menuliskannya ke file output
            for record in SeqIO.parse(infile, "fasta"):
                SeqIO.write(record, outfile, "fasta")

print(f"Sekuen dari {len(daftar_fasta)} file telah digabungkan ke dalam {file_output}")


Sekuen dari 5 file telah digabungkan ke dalam 121450036_Rika Ajeng Finatih.fasta


In [ ]:
from Bio import SeqIO

def fasta_ke_genbank(file_fasta_input, file_genbank_output, molecule_type="DNA"):
    try:
        # Membaca file FASTA
        rekaman = list(SeqIO.parse(file_fasta_input, "fasta"))

        # Membuat daftar baru dari rekaman dengan format GenBank
        rekaman_genbank = [SeqIO.SeqRecord(
            seq=rekaman.seq,
            id=rekaman.id,
            name=rekaman.name,
            description=rekaman.description,
            annotations={"molecule_type": molecule_type}  # Menambahkan tipe molekul
        ) for rekaman in rekaman]

        # Menulis rekaman GenBank ke file output
        SeqIO.write(rekaman_genbank, file_genbank_output, "genbank")

        print(f"Berhasil mengonversi {file_fasta_input} menjadi {file_genbank_output}")
    except FileNotFoundError:
        print(f"Error: File tidak ditemukan - {file_fasta_input}")
    except Exception as e:
        print(f"Terjadi kesalahan: {e}")

# Contoh penggunaan
file_fasta_input = "121450036_Rika Ajeng Finatih.fasta"
file_genbank_output = "121450036_Rika Ajeng Finatih.gb"
fasta_ke_genbank(file_fasta_input, file_genbank_output)


Berhasil mengonversi 121450036_Rika Ajeng Finatih.fasta menjadi 121450036_Rika Ajeng Finatih.gb


Kode ini berhasil mengunduh dan menggabungkan sekuens dari dua format file (FASTA dan GenBank) menjadi satu file dalam format FASTA. Hal ini bermanfaat untuk analisis bioinformatika yang memerlukan data sekuens dari berbagai sumber dalam format yang konsisten. Dengan menggunakan library Biopython, pengolahan file sekuens menjadi lebih mudah dan efisien.